mkdocs/livereload/__init__.py

```diff
@@ -0,0 +1,217 @@
+                try:
+                    return super().translate_path(path)
+                finally:
+                    os.chdir(old_cwd)
+
+        return super().translate_path(path)
+
+    def send_head(self):
+        epoch = self.server._visible_epoch
+
+        file = super().send_head()
+        if file and getattr(file, "name", "").endswith(".html"):
+            try:
+                content = file.read()
+            finally:
+                file.close()
+            try:
+                body_end = content.rindex(b"</body>")
+            except ValueError:
+                body_end = len(content)
+            # The page will reload if the livereload poller returns a newer epoch than what it knows.
+            # The other timestamp becomes just a unique identifier for the initiating page.
+            file = io.BytesIO(
+                b'%b<script src="/js/livereload.js"></script><script>livereload(%d, %d);</script>%b'
+                % (content[:body_end], epoch, _timestamp(), content[body_end:])
+            )
+        return file
+
+    def do_GET(self):
+        m = re.fullmatch(r"/livereload/([0-9]+)/([0-9]+)", self.path)
+        if m:
+            return self._do_poll_response(epoch=int(m[1]), request_id=int(m[2]))
+
+        self.server.wait_for_build()  # Otherwise we may be looking at a half-built site.
+        return super().do_GET()
+
+    def _do_poll_response(self, epoch, request_id):
+        self.send_response(http.HTTPStatus.OK)
+        self.send_header("Content-type", "text/plain")
+        self.end_headers()
+        self._log_poll_request(self.headers.get("referer"), epoch, request_id)
+        # Stall the browser, respond as soon as there's something new.
+        # If there's not, respond anyway after a minute.
+        epoch = self.server.wait_for_epoch(epoch, timeout=self.server.poll_response_timeout)
+        self.wfile.write(b"%d" % epoch)
+
+    @classmethod
+    @functools.lru_cache()  # "Cache" to not repeat the same message for the same browser tab.
+    def _log_poll_request(cls, url, epoch, request_id):
+        log.info(f"Browser connected: {url}")
+
+    def log_message(self, format, *args):
+        log.debug(format, *args)
+
+    def log_error(self, format, *args):
+        log.warning('"%s" ' + format, self.requestline, *args)
+
+    def send_error(self, code, message=None, *args, **kwargs):
+        try:
+            error_content = self.server.error_handler(code)
+        except Exception:
+            log.exception("Failed to render an error message")
+        else:
+            if error_content is not None:
+                self.log_error("code %d", code)
+                self.send_response(code, message)
+                self.end_headers()
+                self.wfile.write(error_content)
+                return
+
+        return super().send_error(code, message, *args, **kwargs)
+
+
+LiveReloadRequestHandler.extensions_map[".js"] = "application/javascript"
```

Comment:
Right. What prompted me to add this is that the new livereload itself would serve the wrong MIME type and break.

The whole point of this *is* to prevent OS level configs from affecting this. Can't be made worse. Though of course I know you mean it's about the "floodgate" for all the other types.

I think indeed, specifying it's just for MkDocs-official types should be fine.

Regarding other common types like JSON that you mention: I think it would be good to just add unittests for all these types (without specifying them in this mapping).